universe/supplyverifier/manager.go

```diff
@@ -0,0 +1,399 @@
+// group does not exist, it will be created and started.
+func (m *Manager) CanHandle(assetSpec asset.Specifier,
+	msg msgmux.PeerMsg) (bool, error) {
+
+	sm, err := m.fetchStateMachine(assetSpec)
+	if err != nil {
+		return false, fmt.Errorf("unable to get or create state "+
+			"machine: %w", err)
+	}
+
+	return sm.CanHandle(msg), nil
+}
+
+// Name returns the name of the state machine associated with the given asset
+// specifier. If a state machine for the asset group does not exist, it will be
+// created and started.
+func (m *Manager) Name(assetSpec asset.Specifier) (string, error) {
+	sm, err := m.fetchStateMachine(assetSpec)
+	if err != nil {
+		return "", fmt.Errorf("unable to get or create state "+
+			"machine: %w", err)
+	}
+
+	return sm.Name(), nil
+}
+
+// SendMessage sends a message to the state machine associated with the given
+// asset specifier. If a state machine for the asset group does not exist, it
+// will be created and started.
+func (m *Manager) SendMessage(ctx context.Context,
+	assetSpec asset.Specifier, msg msgmux.PeerMsg) (bool, error) {
+
+	sm, err := m.fetchStateMachine(assetSpec)
+	if err != nil {
+		return false, fmt.Errorf("unable to get or create state "+
+			"machine: %w", err)
+	}
+
+	return sm.SendMessage(ctx, msg), nil
+}
+
+// CurrentState returns the current state of the state machine associated with
+// the given asset specifier. If a state machine for the asset group does not
+// exist, it will be created and started.
+func (m *Manager) CurrentState(assetSpec asset.Specifier) (
+	protofsm.State[Event, *Environment], error) {
+
+	sm, err := m.fetchStateMachine(assetSpec)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get or create state "+
+			"machine: %w", err)
+	}
+
+	return sm.CurrentState()
+}
+
+// RegisterStateEvents registers a state event subscriber with the state machine
+// associated with the given asset specifier. If a state machine for the asset
+// group does not exist, it will be created and started.
+func (m *Manager) RegisterStateEvents(
+	assetSpec asset.Specifier) (StateSub, error) {
+
+	sm, err := m.fetchStateMachine(assetSpec)
+	if err != nil {
+		return nil, fmt.Errorf("unable to get or create state "+
+			"machine: %w", err)
+	}
+
+	return sm.RegisterStateEvents(), nil
+}
+
+// RemoveStateSub removes a state event subscriber from the state machine
+// associated with the given asset specifier. If a state machine for the asset
+// group does not exist, it will be created and started.
+func (m *Manager) RemoveStateSub(assetSpec asset.Specifier,
+	sub StateSub) error {
+
+	sm, err := m.fetchStateMachine(assetSpec)
+	if err != nil {
+		return fmt.Errorf("unable to get or create state "+
+			"machine: %w", err)
+	}
+
+	sm.RemoveStateSub(sub)
+
+	return nil
+}
+
+// stateMachineCache is a thread-safe cache mapping an asset group's public key
+// to its supply verifier state machine.
+type stateMachineCache struct {
```

Comment:
Can we extract this into a separate package so we don't have to re-defined here? Looks to be identical to the one in `supplycommit` (but probably needs to be parametrized since it's a different state machine struct that's cached).